src/field_theory/separable_degree.lean

```diff
@@ -0,0 +1,183 @@
+def separable_degree : ℕ := nat_degree (separable_contraction q hf)
+
+/-- The separable degree divides the degree, in function of the exponential characteristic of F. -/
+lemma separable_degree_dvd_degree' : ∃ m : ℕ, (separable_degree q hf) * (q ^ m) = f.nat_degree :=
+begin
+  cases (classical.some_spec hf).2 with m hm,
+  use m,
+  let g := classical.some hf,
+  have deg_g : g.nat_degree = separable_degree q hf, by refl,
+  rw [←deg_g, ←hm, nat_degree_expand (q^m) g, mul_comm],
+end
+
+/-- The separable degree divides the degree. -/
+lemma separable_degree_dvd_degree :
+  (separable_degree q hf) ∣ f.nat_degree :=
+let ⟨a, ha⟩ := separable_degree_dvd_degree' q hf in dvd.intro (q ^ a) ha
+
+/-- In exponential characteristic one, the separable degree equals the degree. -/
+lemma separable_degree_eq_degree {f : polynomial F}
+  (hf : has_separable_contraction 1 f) : (separable_degree 1 hf) = f.nat_degree :=
+exists.elim (separable_degree_dvd_degree' 1 hf) (λ a, λ ha, by rw [←ha, one_pow a, mul_one])
+
+lemma exists_eq_add_nonneg_of_lt (m m' : ℕ) (h : m < m') : ∃ s : ℕ, 0 < s ∧ m' = m + s :=
+let ⟨k, hk⟩ := nat.exists_eq_add_of_lt h in ⟨k.succ, k.zero_lt_succ, hk⟩
+
+end comm_semiring
+
+section field
+
+variables {F : Type} [field F]
+  (q : ℕ)
+  {f : polynomial F} (hf : has_separable_contraction q f)
+
+/-- Every irreducible polynomial can be contracted to a separable polynomial.
+https://stacks.math.columbia.edu/tag/09H0 -/
+lemma irreducible_has_separable_contraction (q : ℕ) [hF : exp_char F q]
+  (f : polynomial F) [irred : irreducible f] (fn : f ≠ 0) : has_separable_contraction q f :=
+begin
+  casesI hF,
+  { use f,
+    exact ⟨irreducible.separable irred, ⟨0, by rw [pow_zero, expand_one]⟩⟩ },
+  { haveI qp : fact (nat.prime q) := fact_iff.mpr hF_hprime,
+    rcases exists_separable_of_irreducible q irred fn with ⟨n, g, hg⟩,
+    use g,
+    exact ⟨hg.1, ⟨n, hg.2⟩⟩, }
+end
+
+/-- A helper lemma: if two expansions (along the positive characteristic) of two polynomials `g` and
+`g'` agree, then either their degrees are the same, or one of them is not separable. -/
+lemma contraction_degree_eq_aux [hq : fact q.prime] [hF : char_p F q]
+  (g g' : polynomial F) (m m' : ℕ)
+  (h_expand : expand F (q^m) g = expand F (q^m') g')
+  (h : m < m') :
+  (g.nat_degree =  g'.nat_degree) ∨ ¬g.separable :=
+begin
+  suffices : g.separable → g.nat_degree = g'.nat_degree, by begin
+    apply or_iff_not_imp_right.mpr,
+    rw not_not,
+    exact this,
+  end,
+
+  cases exists_eq_add_nonneg_of_lt m m' h with s hs,
+  intro g_sep,
+
+  rw [hs.2, pow_add, expand_mul] at h_expand,
+
+  have r := expand_injective (pow_pos hq.1.pos m) h_expand,
+  rw r at g_sep,
+
+  cases (is_unit_or_eq_zero_of_separable_expand q s g_sep) with g'_is_unit s_zero,
+  { rw [r, nat_degree_expand (q^s) g',
+      nat_degree_eq_of_degree_eq_some (degree_eq_zero_of_is_unit g'_is_unit), zero_mul ] },
+  { apply false.elim,
+    exact (ne_of_lt hs.1).symm s_zero },
```

Comment:
I don't think the new lemma helps much here:
```suggestion
  obtain ⟨s, rfl⟩ := nat.exists_eq_add_of_lt h,

  rw [add_assoc, pow_add, expand_mul] at h_expand,

  rw expand_injective (pow_pos hq.1.pos m) h_expand,
  intro g_sep,

  have := (is_unit_or_eq_zero_of_separable_expand q s.succ g_sep).resolve_right s.succ_ne_zero,
  rw [nat_degree_expand (q^s.succ) g',
      nat_degree_eq_of_degree_eq_some (degree_eq_zero_of_is_unit this), zero_mul]
```

Reply:
Just copying this suggestion (and replacing g_sep by hg) did not work, I got a timeout. I have therefore replaced it with something similar (albeit slightly less elegant). Is this OK for you?